From 60a11c85e7c4792c797c7a4fcce9fdfa37935310 Mon Sep 17 00:00:00 2001
From: Gerion Entrup <gerion.entrup@flump.de>
Date: Wed, 13 Jan 2021 19:48:57 +0100
Subject: [PATCH] Proto*: fix Error macro conflict

utils/Logger.h contains a preprocessor macro Error() for easy logging.
However, the upstream protobuf/io/coded_stream.h also defines a private
function Error() in one of its classes that conflicts with this macro.

Thus, unfortunate include orders result in a build error. This commit
reorders the includes to prevent the error. It also defines an extra
`#undef Error` before including upstream protobuf to make the problem
more visible (a compiler error would not point to an obscure place
somewhere in protobuf but to the usage of the Error() macro within
Hyperion itself).

Fixes: #882
---
 libsrc/protoserver/ProtoClientConnection.cpp | 7 ++++---
 libsrc/protoserver/ProtoClientConnection.h   | 8 +++++---
 libsrc/protoserver/ProtoServer.cpp           | 2 +-
 3 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/libsrc/protoserver/ProtoClientConnection.cpp b/libsrc/protoserver/ProtoClientConnection.cpp
index e8bf1d918..051ca3e71 100644
--- a/libsrc/protoserver/ProtoClientConnection.cpp
+++ b/libsrc/protoserver/ProtoClientConnection.cpp
@@ -1,12 +1,13 @@
-// project includes
-#include "ProtoClientConnection.h"
-
 // qt
 #include <QTcpSocket>
 #include <QHostAddress>
 #include <QTimer>
 #include <QRgb>
 
+// project includes
+#include "ProtoClientConnection.h"
+
+
 // TODO Remove this class if third-party apps have been migrated (eg. Hyperion Android Grabber, Windows Screen grabber etc.)
 
 ProtoClientConnection::ProtoClientConnection(QTcpSocket* socket, int timeout, QObject *parent)
diff --git a/libsrc/protoserver/ProtoClientConnection.h b/libsrc/protoserver/ProtoClientConnection.h
index 6d8ce3780..e08658dcf 100644
--- a/libsrc/protoserver/ProtoClientConnection.h
+++ b/libsrc/protoserver/ProtoClientConnection.h
@@ -1,14 +1,16 @@
 #pragma once
 
+// protobuffer PROTO
+// protobuf defines an Error() function itself, so undef it here
+#undef Error
+#include "message.pb.h"
+
 // util
 #include <utils/Logger.h>
 #include <utils/Image.h>
 #include <utils/ColorRgb.h>
 #include <utils/Components.h>
 
-// protobuffer PROTO
-#include "message.pb.h"
-
 class QTcpSocket;
 class QTimer;
 
diff --git a/libsrc/protoserver/ProtoServer.cpp b/libsrc/protoserver/ProtoServer.cpp
index 7bfc2954d..fb45ae60c 100644
--- a/libsrc/protoserver/ProtoServer.cpp
+++ b/libsrc/protoserver/ProtoServer.cpp
@@ -1,5 +1,5 @@
-#include <protoserver/ProtoServer.h>
 #include "ProtoClientConnection.h"
+#include <protoserver/ProtoServer.h>
 
 // util
 #include <utils/NetOrigin.h>
